rmf_reservation_node/src/main.cpp

```diff
@@ -0,0 +1,590 @@
+      allocation_pub_->publish(allocation);
+      return;
+    }
+
+    // If we can't proceed immediately add the ticket to a queue.
+    RCLCPP_INFO(
+      this->get_logger(), "Could not immediately service %lu, enqueing. Locations in ticket were:",
+      request->ticket.ticket_id);
+    for (std::size_t i = 0; i < requests_[request->ticket.ticket_id].size();
+      i++)
+    {
+      RCLCPP_INFO(this->get_logger(), "\t- %s",
+        requests_[request->ticket.ticket_id][i].location.c_str());
+    }
+    queue_manager_.add_to_queue(request->ticket.ticket_id, location_names);
+
+    // Allocate a waitpoint by preference as given by Fleet Adapter
+    std::vector<LocationReq> wait_points;
+    auto cost = 0.0;
+    for (auto waitpoint_name: request->wait_points)
+    {
+
+      LocationReq request{
+        waitpoint_name,
+        cost
+      };
+      wait_points.push_back(request);
+      cost += 1.0;
+    }
+
+    waitpoints_[request->ticket.ticket_id] = wait_points;
+    auto waitpoint_result = current_state_.allocate_lowest_cost_free_spot(
+      wait_points,
+      request->ticket.ticket_id);
+    if (waitpoint_result.has_value())
+    {
+      rmf_reservation_msgs::msg::ReservationAllocation allocation;
+      allocation.ticket = ticket_store_.get_existing_ticket(
+        request->ticket.ticket_id);
+      allocation.instruction_type =
+        rmf_reservation_msgs::msg::ReservationAllocation::WAIT_PERMANENTLY;
+      allocation.satisfies_alternative = waitpoint_result.value();
+      allocation.resource = wait_points[waitpoint_result.value()].location;
+      RCLCPP_INFO(this->get_logger(), "Allocating %s as waitpoint to %s",
+        allocation.resource.c_str(),
+        ticket_store_.debug_ticket(request->ticket.ticket_id).c_str());
+      allocation_pub_->publish(allocation);
+    }
+    else
+    {
+      RCLCPP_ERROR(
+        this->get_logger(), "Could not allocate a waiting point for robots from %lu waitpoints",
+        wait_points.size());
+    }
+  }
+
+  void release(
+    const rmf_reservation_msgs::msg::ReleaseRequest::ConstSharedPtr& request)
+  {
+    RCLCPP_INFO(
+      this->get_logger(), "Releasing ticket for %s",
+      ticket_store_.debug_ticket(request->ticket.ticket_id).c_str());
+    auto ticket = request->ticket.ticket_id;
+    auto released_location = current_state_.release(ticket);
+    if (!released_location.has_value())
+    {
+      RCLCPP_ERROR(
+        this->get_logger(), "Could not find ticket %lu",
+        request->ticket.ticket_id);
+      return;
+    }
+    RCLCPP_INFO(
+      this->get_logger(), "Released ticket %lu, location %s is now free",
+      request->ticket.ticket_id,
+      released_location->c_str());
+
+    // Traverse waitgraph.
+    while (auto next_ticket =
+      queue_manager_.service_next_in_queue(released_location.value()))
+    {
+      // Release the ticket
+      released_location = current_state_.release(next_ticket.value());
```

Comment:
This triggers the next robot to "release" its waiting spot so that the next robot can come in to the location. I'm not sure `finalize` captures that relation. Maybe `make_space_for_next_robot`. `service_next_in_queue` could be renamed to `notify_next_robot`.

Reply:
Okay I misunderstood the purpose then. I thought `next_ticket` would represent the reservation for the next robot that is being served `released_location`, so we wouldn't be releasing that ticket now. But now I understand that `next_ticket` represents the location where the next robot had been waiting.

I wonder if renaming it to something like `previous_waiting_location` would make the logic more clear. It's okay either way, but I suspect if anyone looks over this code in the future they might experience the same confusion that I did.